### ISSUE TITLE
errors in never-included files

### ISSUE THREAD

**Issue by wendt88** [NativeScript/nativescript-dev-webpack#1004]
**Environment**
using `nativescript-vue`:
```
√ Getting NativeScript components versions information...
√ Component nativescript has 6.0.2 version and is up to date.
√ Component tns-core-modules has 6.0.1 version and is up to date.
√ Component tns-android has 6.0.0 version and is up to date.
√ Component tns-ios has 6.0.1 version and is up to date.
```
 
packages:
```
"dependencies": {
    "@mdi/font": "^3.7.95",
    "@vue/devtools": "^5.1.1",
    "axios": "^0.19.0",
    "base-64": "^0.1.0",
    "lodash": "^4.17.11",
    "md5": "^2.2.1",
    "mime": "^2.4.3",
    "moment": "^2.24.0",
    "nativescript-background-http": "^3.4.0",
    "nativescript-cfalert-dialog": "^1.0.15",
    "nativescript-drop-down": "^5.0.2",
    "nativescript-exit": "^1.0.1",
    "nativescript-feedback": "^1.3.9",
    "nativescript-geolocation": "^5.1.0",
    "nativescript-google-maps-sdk": "^2.7.0",
    "nativescript-imagepicker": "^6.2.0",
    "nativescript-mediafilepicker": "^2.0.16",
    "nativescript-menu": "^1.1.3",
    "nativescript-nativemediapicker": "^1.0.1",
    "nativescript-orientation": "^2.2.1",
    "nativescript-secure-storage": "^2.4.0",
    "nativescript-socketio": "^3.3.1",
    "nativescript-theme-core": "^1.0.6",
    "nativescript-toasty": "^2.0.0",
    "nativescript-ui-autocomplete": "5.0.0",
    "nativescript-ui-listview": "7.0.0",
    "nativescript-ui-sidedrawer": "7.0.0",
    "nativescript-videoplayer": "^4.2.1",
    "nativescript-vue": "^2.2.2",
    "nativescript-vue-devtools": "^1.2.0",
    "tns-core-modules": "6.0.1",
    "vue-async-computed": "^3.7.0",
    "vue-i18n": "^8.11.2",
    "vuex": "^3.1.1"
},
"devDependencies": {
    "@babel/core": "^7.5.0",
    "@babel/preset-env": "7.3.1",
    "babel-eslint": "^10.0.2",
    "babel-loader": "^8.0.6",
    "babel-traverse": "6.26.0",
    "babel-types": "6.26.0",
    "babylon": "6.18.0",
    "eslint": "^6.0.1",
    "eslint-import-resolver-webpack": "^0.11.1",
    "eslint-plugin-standard": "^4.0.0",
    "eslint-plugin-vue": "^5.2.3",
    "intl": "^1.2.5",
    "lazy": "1.0.11",
    "nativescript-dev-webpack": "1.0.1",
    "nativescript-vue-template-compiler": "^2.2.2",
    "net": "^1.0.2",
    "node-sass": "^4.12.0",
    "stylelint": "^10.0.1",
    "stylelint-config-standard": "^18.3.0",
    "stylelint-scss": "^3.9.0",
    "vue-loader": "^15.7.0",
    "webpack": "^4.35.2",
    "webpack-bundle-analyzer": "^3.3.2",
    "webpack-cli": "^3.3.5"
}
```
 

**Describe the bug**
I use a own submodule (git), which i use to an other project. there are some unused files for my mobile application inside.
After upgrade to nativescript 6, and `tns migrate` i get some errors of missing packages like
```
ERROR in <my submodule file>
Module not found: Error: Can't resolve 'bcrypt' in '<my submodule>'
 @ <my submodule file> 9:15-32
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in <my submodule test file>
Module not found: Error: Can't resolve 'tape' in '<my submodule>'
 @ <my submodule test file> 9:15-32
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js
```

those files gets never included in my repo (like test files).
so I tried to install the missing packages like `bcrypt` and `tape`, but so I get a few other errors during build from those packages:

```
ERROR in ../node_modules/node-gyp/lib/node-gyp.js
Module not found: Error: Can't resolve '../package' in 'C:\Projects\framework2-vue-app\node_modules\node-gyp\lib'
 @ ../node_modules/node-gyp/lib/node-gyp.js 53:16-37
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-pre-gyp/lib/unpublish.js
Module not found: Error: Can't resolve 'aws-sdk' in 'C:\Projects\framework2-vue-app\node_modules\node-pre-gyp\lib'
 @ ../node_modules/node-pre-gyp/lib/unpublish.js 23:12-30
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-pre-gyp/lib/publish.js
Module not found: Error: Can't resolve 'aws-sdk' in 'C:\Projects\framework2-vue-app\node_modules\node-pre-gyp\lib'
 @ ../node_modules/node-pre-gyp/lib/publish.js 27:12-30
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-pre-gyp/lib/info.js
Module not found: Error: Can't resolve 'aws-sdk' in 'C:\Projects\framework2-vue-app\node_modules\node-pre-gyp\lib'
 @ ../node_modules/node-pre-gyp/lib/info.js 19:12-30
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/detect-libc/lib/detect-libc.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\detect-libc\lib'
 @ ../node_modules/detect-libc/lib/detect-libc.js 5:16-40
 @ ../node_modules/node-pre-gyp/lib/util/versioning.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-gyp/lib/node-gyp.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-gyp\lib'
 @ ../node_modules/node-gyp/lib/node-gyp.js 13:20-44
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-gyp/lib/find-vs2017.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-gyp\lib'
 @ ../node_modules/node-gyp/lib/find-vs2017.js 2:15-39
 @ ../node_modules/node-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-gyp/lib/node-gyp.js
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-gyp/lib/build.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-gyp\lib'
 @ ../node_modules/node-gyp/lib/build.js 12:11-35
 @ ../node_modules/node-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-gyp/lib/node-gyp.js
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-gyp/lib/configure.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-gyp\lib'
 @ ../node_modules/node-gyp/lib/configure.js 19:9-33
 @ ../node_modules/node-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-gyp/lib/node-gyp.js
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-pre-gyp/lib/testbinary.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-pre-gyp\lib'
 @ ../node_modules/node-pre-gyp/lib/testbinary.js 12:9-33
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/node-pre-gyp/lib/util/compile.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\node-pre-gyp\lib\util'
 @ ../node_modules/node-pre-gyp/lib/util/compile.js 12:9-33
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/osenv/osenv.js
Module not found: Error: Can't resolve 'child_process' in 'C:\Projects\framework2-vue-app\node_modules\osenv'
 @ ../node_modules/osenv/osenv.js 5:11-35
 @ ../node_modules/node-gyp/lib/install.js
 @ ../node_modules/node-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-gyp/lib/node-gyp.js
 @ ../node_modules/node-pre-gyp/lib/util/compile.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/forever-agent/index.js
Module not found: Error: Can't resolve 'tls' in 'C:\Projects\framework2-vue-app\node_modules\forever-agent'
 @ ../node_modules/forever-agent/index.js 7:10-24
 @ ../node_modules/request/request.js
 @ ../node_modules/request/index.js
 @ ../node_modules/node-pre-gyp/lib/install.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js

ERROR in ../node_modules/tunnel-agent/index.js
Module not found: Error: Can't resolve 'tls' in 'C:\Projects\framework2-vue-app\node_modules\tunnel-agent'
 @ ../node_modules/tunnel-agent/index.js 4:10-24
 @ ../node_modules/request/lib/tunnel.js
 @ ../node_modules/request/request.js
 @ ../node_modules/request/index.js
 @ ../node_modules/node-pre-gyp/lib/install.js
 @ ../node_modules/node-pre-gyp/lib sync ^\.\/.*$
 @ ../node_modules/node-pre-gyp/lib/node-pre-gyp.js
 @ ../node_modules/bcrypt/bcrypt.js
 @ ./api/sdk/tools/crypto.js
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./main.js
```

**Additional context**
why all files in `./app` get built?
and how can I ignore some files from the build-process?

**Comment by DimitarTachev** [comment]
Hi @wendt88,

We include all of your app files in the application bundle. In order to do that, we include all files based on your **app entry** (`main.js`).

In other words, it seems that you have a `node_modules` folder somewhere in your app folder.

In order to avoid getting these errors, you could move your `node_modules` (or maybe the submodule if it contains another `node_modules`) at your workspace root, while your app entry should be located in some `<workspace-root>/app` or `<workspace-root>/src` folder. In this way, you will not get any `node_modules` as we will include only the files from your `<workspace-root>/app` folder.

If you don't want to change your app structure, you could ignore the `node_modules` by modifying your `webpack.config.js` and add an additional `"node_modules"` string to the `ignoredFiles` array of the `bundle-config-loader`. However, the Webpack build process will still go through all `node_modules` and your Webpack compilation time will be slower than not having `node_modules` there at all.

If you have some issues with your app structure setup, you could take a look at the recommended structure by executing `tns create newApp --vue`.

**Comment by wendt88** [comment]
[this project](https://play.nativescript.org/?template=play-vue&id=AeUUXT) works in the playground, but if i download it and run `tns prepare android --bundle`, i get the same error:
```
ERROR in ./helpers/never-included.js
Module not found: Error: Can't resolve 'bcrypt' in '<LOCATION>\app\helpers'
 @ ./helpers/never-included.js 1:13-30
 @ . sync (?<!\bApp_Resources\b.*)\.(xml|css|js|(?<!\.d\.)ts|(?<!\b_[\w-]*\.)scss)$
 @ ./app.js
```

**Comment by DimitarTachev** [comment]
Hi @wendt88,

The Playground has [a limited functionality](https://www.nativescript.org/blog/architecture-of-the-nativescript-playground) and it's just syncing a few files to an already build application on your phone. It does not have `prepare` and `build` phases.

When you download the app and execute `tns run`, you are preparing, building and testing your actual app. In this case, you are getting the above-mentioned error because starting from `nativescript-dev-webpack@1.0.0`, all js files from your app folder are automatically included in the app during the `prepare` phase (take a look at [this PR](https://github.com/NativeScript/nativescript-dev-webpack/pull/966)).

In other words, this is the expected behavior for NativeScript Core applications (I believe it will not be reproducible in NativeScript-Angular apps) because starting from NativeScript 6.0, we are automatically including all app files instead of requiring the old `-page` or `-root` convention which was causing a lot of strange issues and confusion.

If you want the Playground experience locally, you could always use `tns preview`.

**Comment by wendt88** [comment]
ok, so EVERY file inside `./app` gets webpacked?
Than would be nice to give in some war the possibility to extend the current webpack.config.js in some way to not change it after every update

**Comment by DimitarTachev** [comment]
@wendt88,

Yes, every file from the app folder gets Webpacked.

You could exclude files or folders by providing additional values in the [`ignoredFiles` option](https://github.com/NativeScript/nativescript-dev-webpack/blob/469c4cbf381529c28b70afd73c4bed0b6384260b/bundle-config-loader.ts#L22-L24) of the `bundle-config-loader` in your `webpack.config.js`.

The default config curently looks like the following one:
```JavaScript
{
    loader: "nativescript-dev-webpack/bundle-config-loader",
    options: {
        loadCss: !snapshot, // load the application css if in debug mode
        unitTesting,
        appFullPath,
        projectRoot,
        ignoredFiles: nsWebpack.getUserDefinedEntries(entries, platform)
    }
},
```

You could find examples of excluding a file and a folder in the code below:
```JavaScript
const ignoredFiles = nsWebpack.getUserDefinedEntries(entries, platform);
ignoredFiles.push("myIgnoredFolder");
ignoredFiles.push("onlyTheNameOfTheIgnoredFileWithoutExtension");
// a combination folder + file in order to avoid excluding multiple files with the same name in different folders
ignoredFiles.push("theFolder/onlyTheNameOfTheIgnoredFileWithoutExtension");
...
{
    loader: "nativescript-dev-webpack/bundle-config-loader",
    options: {
        loadCss: !snapshot, // load the application css if in debug mode
        unitTesting,
        appFullPath,
        projectRoot,
        ignoredFiles
    }
},
...
```

**Comment by tkrevh** [comment]
We ran into the same issue after upgrading to NS6, because we're sharing same codebase between web and nativescript app. After some investigation, we came up with this solution (in `vue.config.js`) - we had to exclude main.js and router.js:

`  chainWebpack: config => {
    if (process.env.VUE_CLI_TARGET == 'nativescript') {
      config.module
        .rule('native-loaders')
        .uses.get('nativescript-dev-webpack/bundle-config-loader')
        .tap(options => {
          options.ignoredFiles = ['router', 'main']
          return options
        })
    }
  },
`
